DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp

```diff
@@ -0,0 +1,552 @@
+  // fast kernel relaunch (see the README). If this number is set to high,
+  // then the first kernel launched finishes before we are done launching all
+  // the kernels and therefore throughput is decreased.
+  size_t inflight_kernels = 2;
+
+  // parse the command line arguments
+  for (int i = 1; i < argc; i++) {
+    std::string arg(argv[i]);
+
+    if (arg == "--help" || arg == "-h") {
+      std::cout << "USAGE: "
+                << "./simple_host_streaming "
+                << "[--chunks=<int>] "
+                << "[--chunk_count=<int>] "
+                << "[--inflight_kernels=<int>] "
+                << "[--iterations=<int>]\n";
+      return 0;
+    } else {
+      std::string str_after_equals = arg.substr(arg.find("=") + 1);
+
+      if (arg.find("--chunks=") == 0) {
+        chunks = atoi(str_after_equals.c_str());
+      } else if (arg.find("--chunk_count=") == 0) {
+        chunk_count = atoi(str_after_equals.c_str());
+      } else if (arg.find("--inflight_kernels=") == 0) {
+        inflight_kernels = atoi(str_after_equals.c_str());
+      } else if (arg.find("--iterations=") == 0) {
+        iterations = std::max(2, atoi(str_after_equals.c_str()) + 1);
+      } else {
+        std::cout << "WARNING: ignoring unknown argument '" << arg << "'\n";
+      }
+    }
+  }
+
+  // check the chunks
+  if (chunks <= 0) {
+    std::cerr << "ERROR: 'chunks' must be greater than 0\n";
+    return 1;
+  }
+
+  // check the chunk size
+  if (chunk_count <= 0) {
+    std::cerr << "ERROR: 'chunk_count' must be greater than 0\n";
+    return 1;
+  }
+
+  // check inflight_kernels
+  if (inflight_kernels <= 0) {
+    std::cerr << "ERROR: 'inflight_kernels' must be positive\n";
+    return 1;
+  }
+
+  // check the number of iterations
+  if (iterations <= 0) {
+    std::cerr << "ERROR: 'iterations' must be positive\n";
+    return 1;
+  }
+
+  // compute the total number of elements
+  size_t total_count = chunks * chunk_count;
+
+  std::cout << "# Chunks:             " << chunks << "\n";
+  std::cout << "Chunk count:          " << chunk_count << "\n";
+  std::cout << "Total count:          " << total_count << "\n";
+  std::cout << "Iterations:           " << iterations-1 << "\n";
+  std::cout << "\n";
+
+  bool passed = true;
+
+  try {
+    // device selector
+#if defined(FPGA_EMULATOR)
+    INTEL::fpga_emulator_selector selector;
+#else
+    INTEL::fpga_selector selector;
+#endif
+
+    // queue properties to enable profiling
+    auto prop_list = property_list{ property::queue::enable_profiling() };
+
+    // create the device queue
+    queue q(selector, dpc_common::exception_handler, prop_list);
+
+    // make sure the device supports USM host allocations
+    device d = q.get_device();
+    if (!d.get_info<info::device::usm_host_allocations>()) {
+      std::cerr << "ERROR: The selected device does not support USM host"
+                << " allocations\n";
+      return 1;
```

Comment:
It is not clear why sometimes you `return 1` and sometimes you `std::terminate()`.

Reply:
ctrl-v ... :)